native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/QueryProcessor.java

```diff
@@ -66,61 +74,113 @@ private QueryProcessor() {
     public static BStream nativeQuery(
             BObject client, Object paramSQLString,
             Object recordType,
+            AbstractStatementParameterProcessor statementParameterProcessor,
+                AbstractResultParameterProcessor resultParameterProcessor) {
+        ResultSet resultSet = null;
+        try {
+            resultSet = initQuery(client, paramSQLString, statementParameterProcessor);
+            RecordType streamConstraint = (RecordType) ((BTypedesc) recordType).getDescribingType();
+            List<ColumnDefinition> columnDefinitions = Utils.getColumnDefinitions(resultSet, streamConstraint);
+            return ValueCreator.createStreamValue(TypeCreator.createStreamType(streamConstraint,
+                    PredefinedTypes.TYPE_NULL), resultParameterProcessor
+                    .createRecordIterator(resultSet, statement, connection, columnDefinitions, streamConstraint));
+        } catch (SQLException e) {
+            Utils.closeResources(trxResourceManager, resultSet, statement, connection);
+            BError errorValue = ErrorGenerator.getSQLDatabaseError(e,
+                    "Error while executing SQL query: " + sqlQuery + ". ");
+            return ValueCreator.createStreamValue(TypeCreator.createStreamType(Utils.getDefaultStreamConstraint(),
+                    PredefinedTypes.TYPE_NULL), createRecordIterator(errorValue));
+        } catch (ApplicationError applicationError) {
+            Utils.closeResources(trxResourceManager, resultSet, statement, connection);
+            BError errorValue = ErrorGenerator.getSQLApplicationError(applicationError.getMessage());
+            return getErrorStream(recordType, errorValue);
+        } catch (Throwable e) {
+            Utils.closeResources(trxResourceManager, resultSet, statement, connection);
```

Comment:
We can move the close resources to finally block.

Reply:
Done is 795c88c